### ISSUE TITLE
Hadoop indexing time increased from 40 mins to 4 hours after upgrading to 0.9.0

### ISSUE THREAD

**Issue by spektom** [druid-io/druid#2906]
All the Hadoop tasks are finished in 40 minutes, except for the last 2 reduce ones, which are stack for the rest of the time.

We've re-ordered dimensions under dimensionSpec so they come from low cardinality to high, as it was stated in the Druid 0.9.0 changelog.

Long running reduce task log is attached.

[stack_reduce.log.gz](https://github.com/druid-io/druid/files/244572/stack_reduce.log.gz)

**Comment by gianm** [comment]
@spektom what is performance like if you order your dimensions alphabetically? This is what 0.8.x would have done, so that'll be an apples to apples comparison.

**Comment by spektom** [comment]
@gianm It's the same - 2 latest jobs are stuck.

**Comment by fjy** [comment]
@spektom Can you provide us more information about this particular problem you are seeing? Can you provide us with a profile of the indexing and also provide some information about the data you have. Are there a lot of nulls in your data?

**Comment by spektom** [comment]
@fjy there are no nulls in the data. 

I'm attaching an indexing spec file for a while (please note that this is a template, and some uppercase symbols are replaced with actual dates, like INTERVAL_START):
[spec.zip](https://github.com/druid-io/druid/files/257849/spec.zip)

I'm also attaching segments metadata:
[metadata.zip](https://github.com/druid-io/druid/files/257872/metadata.zip)

Please advice what other information would be helpful.

**Comment by fjy** [comment]
@spektom can you profile the indexing job?

**Comment by spektom** [comment]
@fjy I will try, and report back.

**Comment by du00cs** [comment]
@fjy @spektom I have tracked for this (I have trouble on it) for days and found a way to speed up. Simplest way to solve it is rather simple: change targetPartitionSize to numShards equal to 1 or some reasonable values.

```json
    "tuningConfig": {
      "type": "hadoop",
      "partitionsSpec": {
        "type": "hashed",
        ""numShards": 1,
        "assumeGrouped": true
      }
```

The problem is that if a shard(I think it is nearly equal to say `segment` too) contains rollup records less than `targetPartitionSize`, it will get a `NonShardSpec`. All `NonShardSpec` ones will have the same `shardNum` and same reduce. In my case, rollup records will never reach 5m at all and all records from mappers will flow to the same reducer. And this slow down the indexing, especially when data is large but rollup is small.

To fix this, nonempty shards should not get a `NonShardSpec` and others remain the same. I think It is simple to do that. If I could,  I want to have a try,

**Comment by spektom** [comment]
@du00cs we don't know how many shards it's going to create, since we're rebuilding the data for the whole time period (years) every day. We do this in order to organize events by secondary time dimension.

If I understand correctly, there are some days for which the number of output rows is less than the `targetPartitionSize`, so indexing for these days is performed in one reduce task. This is how the data segments are spread over the timeline:

![screenshot from 2016-05-17 12 58 00](https://cloud.githubusercontent.com/assets/158011/15318611/82e0bff4-1c2f-11e6-9c61-6855052e687c.png)

Thanks for the explanation!

**Comment by du00cs** [comment]
@spektom `numShards` do not mean 1 shard in the end total. It is the number of shards per segment, so feel free to set 1 or 2 or some else.

**Comment by spektom** [comment]
@du00cs It worked! Thanks.

**Comment by drcrallen** [comment]
I'm actually going to re-open this until there's a bit more on the subject. I'm also curious if there was some change in 0.9.0 which modified hashing behavior.

If my understanding of your analysis is correct, you are saying the jobs which were taking crazy long times were all resulting in a none shardspec (meaning only one shard per segment interval), and when you are indexing a large number of segment intervals, all of the none shardspecs go to the same reducer?

**Comment by drcrallen** [comment]
@himanshug do you know of any change which would have caused this behavior in 0.9.0?

**Comment by himanshug** [comment]
that behavior existed pre 0.9.0 release too, i think this manifestation is not from 0.9.0 upgrade but the specific dataset . i believe, 0.8.3 release on same dataset will show same behavior.

**Comment by gianm** [comment]
@himanshug the log from the original comment makes it look like multiple segments are getting made by the same reducer; AFAIK that's not supposed to happen. It should be one segment per reducer, right?

**Comment by himanshug** [comment]
there would be one segment per reducer ( there would be multiple intermediate persists on same reducer though).
we need to see the druid task log really to know number of reducers started but I guess they were more than one.  From the spec printed , there indexer hadoop job should have had 379 reducers.
@spektom  can you share the druid task log?

**Comment by gianm** [comment]
@himanshug i'm looking at the log that @spektom posted in the original comment; it has a lot of "upload complete" messages for different segments. assuming that log is from a single reducer, that seems suspicious.

**Comment by gianm** [comment]
@spektom do you have a copy of the Druid task log? (What you get when you click "log (all)" next to a task in the overlord)

**Comment by spektom** [comment]
@drcrallen Attaching [log.gz](https://github.com/druid-io/druid/files/269607/log.gz)

**Comment by himanshug** [comment]
@spektom that does not seem to have much of task log, generally task log for hadoop index task would contain task definition json, details of the hadoop jobs launched, their counter etc. Did you get it from the overlord console for the task in question?

**Comment by spektom** [comment]
@Himanshu No, I've got it from the deep storage (under indexing-logs/
folder)

**Comment by du00cs** [comment]
@himanshug @gianm I run across the same problem as @spektom. I tracked for some while and finally located it.

In `DetermineHashedPartitionsJob` If a segment (after rollup) has less records than `targetPartition` it will get a `NonShardSpec` and a unique shardNum. This is ok, but when `IndexGeneratorJob` rebuilds the configurations from the former stage, it lost the `shardNum`. Because all the `NonShardSpec` will be put into a `Map<ShardSpec, HadoopyShardSpec> hadoopShardSpecLookup`, and this lead to all `NonShardSpec` shards to have the same `shardNum` and finally actually only one reduce working.

**Comment by nishantmonu51** [comment]
replacing NoneShardSpec with HashBasedNumberedShardSpec works for now since HashBasedNumberedShardSpec does not implement equals method, this seems very fragile way of fixing things that can break in unexpected ways if someone adds an equals method to all the ShardSpecs, 

IMHO, It seems the actual issue is with the `Map<ShardSpec, HadoopyShardSpec> hadoopShardSpecLookup` being used as a data structure in HadoopDruidIndexerConfig which is not accounting for interval for which the shardSpec belong to.

**Comment by nishantmonu51** [comment]
#2985 has the changes i am referring to. in my comment

**Comment by gianm** [comment]
Great catch @du00cs, and thanks @spektom for the report!

**Comment by du00cs** [comment]
Nice quick fix. Thanks @nishantmonu51 , the bug confused me a lot.

**Comment by himanshug** [comment]
@du00cs thanks for investigating and finding the cause. @gian I was looking at the behavior of using none shard spec in the case of less rows than targetPartitionSize which existed since before, thanks for pursuing this.

**Comment by gianm** [comment]
Fixed by #2985
